### ISSUE TITLE
[Feature Request] Support querying joined fields

### ISSUE THREAD

**Issue by Napam** [typesense/typesense#1657]
# Feature request: support querying joined fields
Hi, Typesense has been a great product for me and my team. It is exciting to see the new features in the recent 26.0 release (especially the JOINs). However a missing feature we really wish for is being able to do a "search in joined fields". That is, in the docs at https://typesense.org/docs/26.0/api/joins.html#one-to-one-relation where you do:

```sh
curl "http://localhost:8108/multi_search" -X POST \
        -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
        -d '{
          "searches": [
            {
              "collection": "books",
              "include_fields": "$authors(first_name,last_name)",
              "q": "famous"
            }
          ]
        }'
```
which resulted with
```json
{
  "document": {
    "id": "0",
    "title": "Famous Five",
    "author_id": "0",
    "authors": {
      "first_name": "Enid",
      "last_name": "Blyton"
    }
  }
}
```
we really hoped that you could actually search for the name of the author in that query, like this (see the query text):
```sh
curl "http://localhost:8108/multi_search" -X POST \
        -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
        -d '{
          "searches": [
            {
              "collection": "books",
              "include_fields": "$authors(first_name,last_name)",
              "q": "enid blyton"
            }
          ]
        }'
```
## Our current workaround
Out workaround now is that we have to include the first name and last name of the author in the book entries themselves. Which by itself is not difficult to do. It becomes complex where you change a name of the author, then you have to update all the author's books, but also imagine more complex scenarios where you have to resolve a larger graph of dependencies. Then one needs to actually traverse the graph and update everything based on the relations. Doing so in the application layer is quite slow, and all of this would be solved with being able to query joined fields.

We also tried using synonyms for this, but it didn't really seem to work. Using the book-author example we tried to define a synonym mapping the author name to a book id (of which we have indexed), but that didn't wprl. E.g. we defined a synonym "enid" -> the_book_id, then searching enid returned nothing.

**Comment by abiudmani** [comment]
Hi @Napam I believe this can be achieved using the **"filter_by"** field instead of the **"q"** field in the query to search using the author name from the books collection itself as follows,

**Query:**
curl "http://localhost:8108/multi_search" -X POST \
        -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
        -d '{
          "searches": [
            {
              "q": "*",
              "collection": "books",
              "include_fields": "$authors(first_name,last_name)",
              **"filter_by": "$authors(first_name:enid&&last_name:blyton)"**
            }
          ]
        }'

**Response:**

{
    "facet_counts": [],
    "found": 2,
    "hits": [
        {
            "document": {
                "author_id": "1",
                "authors": {
                    "first_name": "Enid",
                    "last_name": "Blyton"
                },
                "id": "2",
                "title": "The Island of Adventure"
            },
            "highlight": {},
            "highlights": []
        },
        {
            "document": {
                "author_id": "1",
                "authors": {
                    "first_name": "Enid",
                    "last_name": "Blyton"
                },
                "id": "1",
                "title": "The Mountain"
            },
            "highlight": {},
            "highlights": []
        }
    ],
    "out_of": 10,
    "page": 1,
    "request_params": {
        "collection_name": "books",
        "first_q": "*",
        "per_page": 10,
        "q": "*"
    },
    "search_cutoff": false,
    "search_time_ms": 0
}

**Comment by gravy-jones-locker** [comment]
+1 for this feature. it would be incredibly helpful for us to avoid the kind of duplication mentioned by the OP - but I guess there were good reasons for not including it with the join work... @kishorenc are there any existing plans to implement this?

**Comment by kishorenc** [comment]
Yes we have plans to implement this, it's already on our roadmap for join.

**Comment by itajenglish** [comment]
+1 for this

**Comment by JonasTriki** [comment]
Any updates on this @kishorenc?

**Comment by kishorenc** [comment]
Still on our immediate roadmap, but we're yet to pick it up. We have been shipping other join features, so we will get to this soon.

**Comment by scottnei-method** [comment]
@kishorenc Do you know if this will include facet_by on the joined collection too?  Jason said that was being worked on, but I couldn't find a card in GH.
https://typesense-community.slack.com/archives/C01P749MET0/p1733352901584049?thread_ts=1733349082.327899&cid=C01P749MET0

**Comment by itajenglish** [comment]
Hi, any ETA on this release?

**Comment by kishorenc** [comment]
We will be picking this up in the next week or so. But no ETA yet.

**Comment by wian-fixify** [comment]
v28 has been released, but it seems this has not been added as part of it although it is in the roadmap for v28?

**Comment by wian-fixify** [comment]
Moved to Medium term? That's disappointing, I was so excited for v28 because of this :(

**Comment by kishorenc** [comment]
Sorry, this is still on our roadmap. We're knocking off the many related joins-related features one by one. I will update this thread when this is ready.

**Comment by Napam** [comment]
Have I understood it correctly that this feature has not been included in the v29 release?

**Comment by kishorenc** [comment]
Yes, this could not make the cut for the v29 release, but it's very much around the corner, in a future v30 RC build.

**Comment by otarnus-dennemeyer** [comment]
+1 for this feature!

@kishorenc, do you think we we will see that in v30? This would really be a game changer for lot of people here apparently.

**Comment by Trinistar** [comment]
@kishorenc whats the status here?

**Comment by kishorenc** [comment]
Sorry about how long this particular feature has taken to ship. We ran into several complexities with our underlying design  as querying touches on so many features. Unfortunately, we have decided to pause development on this feature for now as we re-evaluate our design and approach.

**Comment by sandstrom** [comment]
Thanks for the update @kishorenc

**Comment by gabrielstuff** [comment]
I think it would be a real game changer to be able to do such a thing. Thanks for all the work @kishorenc and the team !

**Comment by matthewfr-aug** [comment]
Agreed this would be great
